# 3. PyTorch 프로젝트 구조 이해하기

## Reference

- [Pytorch Template](https://github.com/victoresque/pytorch-template)
- [Pytorch Template 2](https://github.com/FrancescoSaverioZuppichini/PyTorch-Deep-Learning-Template)
- [Pytorch Lightning Template](https://github.com/PyTorchLightning/deep-learning-project-template)
- [Pytorch Lightning](https://www.pytorchlightning.ai/)
- [Pytorch Lightning + NNI Boilerplate](https://github.com/davinnovation/pytorch-boilerplate)

<br>

## 3.1 PyTorch Project Template Overview

- 초기 단계에서는 대화식 개발 과정이 유리
  - 학습과정과 디버깅 등 지속적인 확인
- 배포 및 공유 단계에서는 notebook 공유의 어려움
  - 쉬운 재현의 어려움, 실행순서 꼬임
- DL 코드도 하나의 프로그램
  - 개발 용이성 확보와 유지보수 향상 필요


- 코드도 레고블럭 처럼
  - OOP + 모듈 -> 프로젝트

- 다양한 프로젝트 템플릿이 존재
- 사용자 필요에 따라 수정하여 사용
- 실행, 데이터, 모델, 설정, 로깅, 지표, 유틸리티 등 다양한 모듈들을 분리하여 프로젝트 템플릿화

- PyTorch Template 추천 repository
  - [https://github.com/FrancescoSaverioZuppichini/PyTorch-Deep-Learning-Template](https://github.com/FrancescoSaverioZuppichini/PyTorch-Deep-Learning-Template)
  - [https://github.com/PyTorchLightning/deep-learning-project-template](https://github.com/PyTorchLightning/deep-learning-project-template)
  - [https://github.com/victoresque/pytorch-template](https://github.com/victoresque/pytorch-template) <- 이 것이 가장 좋았던 것 같다!

<br>

## 3.2 Module 구성

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<img src='https://drive.google.com/uc?id=1UtFuDmoR5qkoaHiPeJvKSlZGRl29rLqT' width=900/>

<br>

## 3.3 로컬에서 프로젝트 clone 하여 사용하기

- clone 한 프로젝트에서 다음 명령어 실행

```
python new_project.py MyProject
```

- `MyProject` 라는 새로운 프로젝트 생성하겠다는 명령어
- `MyProject` 라는 새로운 디렉토리가 생성됨

```
cd MyProject
```

- `MyProject` 로 이동

```
code .
```

- vscode 를 통해 프로젝트 구조 확인

```
python train.py -c config.json
```

- train 모듈 실행

<br>

## 3.4 colab 에서 프로젝트 clone 하여 사용하기

- 런타임 유형을 GPU 를 선택하고 다음 작업을 실행해야 GPU 를 사용할 수 있다.

### 3.4.1 project clone 및 new project 생성

In [ ]:
# !pip3 install torch
# !pip3 install torchvision

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
%cd /content/drive/MyDrive/Colab Notebooks/boostcamp/04_교육내용/003_PyTorch/03_project template

/content/drive/MyDrive/Colab Notebooks/boostcamp/04_교육내용/003_PyTorch/03_project template


In [ ]:
!git clone https://github.com/victoresque/pytorch-template

Cloning into 'pytorch-template'...
remote: Enumerating objects: 1516, done.
remote: Total 1516 (delta 0), reused 0 (delta 0), pack-reused 1516
Receiving objects: 100% (1516/1516), 288.08 KiB | 1020.00 KiB/s, done.
Resolving deltas: 100% (848/848), done.


In [ ]:
!ls

pytorch-template


In [ ]:
%cd pytorch-template/

/content/drive/My Drive/Colab Notebooks/boostcamp/04_교육내용/003_PyTorch/03_project template/pytorch-template


In [ ]:
!python new_project.py MNIST-example

New project initialized at /content/drive/My Drive/Colab Notebooks/boostcamp/04_교육내용/003_PyTorch/03_project template/pytorch-template/MNIST-example


<br>

### 3.4.2 NGROK 사용

- 코랩 서버에 SSH 로 접속하는 방법
- [ngrok 사이트](https://ngrok.com/) 에서 회원 가입
- "authtoken" 복사 후 붙여넣기

In [ ]:
NGROK_TOKEN = '1wpyqsDeJKZ5jPRPQ307HdSOn6L_3eKMsASrHX5TS4bs3HL2L' # ngrok 토큰
PASSWORD = 'upsatge' # 비밀번호 설정

- `colab-ssh` 설치

In [ ]:
!pip install colab-ssh

- SSH 로 접근할 수 있는 서버 실행

In [ ]:
from colab_ssh import launch_ssh
launch_ssh(NGROK_TOKEN, PASSWORD)

Successfully running 6.tcp.ngrok.io:19805
[Optional] You can also connect with VSCode SSH Remote extension using this configuration:

	Host google_colab_ssh
		HostName 6.tcp.ngrok.io
		User root
		Port 19805
	  


- vscode 에서 "Remote - SSH" Extension 설치
- Action 에서 "Remote-SSH: Add New SSH Host" 선택
- 다음 내용 입력
  - `ssh root@HostName -p Port`
  - `ssh root@6.tcp.ngrok.io -p 19805`
- SSH 저장 위치 선택 (default) 사용
- Action 에서 "Remote-SSH: Connect to Host" 선택
- 위에서 지정한 password 입력
- 터미널에서 `/content` 로 이동 시 colab 과 동일한 경로로 이동한 것을 확인할 수 있다.

- 로컬에서 처럼 train 모듈도 실행해볼 수 있다.
```
python train.py -c config.json
```

- open folder 를 선택해서 파일을 볼 수도 있다.

<br>

### 3.4.3 GoogleDrive 에 저장

```
cp -r /content/저장하고자하는경로 저장할위치
```

```
cp -r /content/pytorch-template /content/drive/MyDrive/workspace
```

<br>

## 3.5 프로젝트 구조 이해

### 3.5.1 `train.py`

- `train.py` 로 시작한다.

```python
args = argparse.ArgumentParser(description='PyTorch Template')
args.add_argument('-c', '--config', default=None, type=str,
                  help='config file path (default: None)')
args.add_argument('-r', '--resume', default=None, type=str,
                  help='path to latest checkpoint (default: None)')
args.add_argument('-d', '--device', default=None, type=str,
                  help='indices of GPUs to enable (default: all)')
...
config = ConfigParser.from_args(args, options)
main(config)
```

- argument 지정
- `-c` : config 지정
  - `parse_config.ConfigParser`
  - `util.read_json` : json to dict

```python
def main(config):
logger = config.get_logger('train') # logger 설정
# setup data_loader instances
data_loader = config.init_obj('data_loader', module_data) # data_loader 불러오기
valid_data_loader = data_loader.split_validation()
# build model architecture, then print to console
model = config.init_obj('arch', module_arch) # 모델 불러오기
logger.info(model)
# prepare for (multi-device) GPU training
device, device_ids = prepare_device(config['n_gpu'])
model = model.to(device)
if len(device_ids) > 1:
    model = torch.nn.DataParallel(model, device_ids=device_ids)
# get function handles of loss and metrics
criterion = getattr(module_loss, config['loss']) # loss 정의
metrics = [getattr(module_metric, met) for met in config['metrics']]
# build optimizer, learning rate scheduler. delete every lines aining lr_scheduler for disabling scheduler
trainable_params = filter(lambda p: p.requires_grad, model.parameters())
optimizer = config.init_obj('optimizer', torch.optim, trainable_params)
lr_scheduler = config.init_obj('lr_scheduler', torch.optim.lr_scheduler, mizer)
# Trainer() : 핵심적인 부분
trainer = Trainer(model, criterion, metrics, optimizer,
                  config=config,
                  device=device,
                  data_loader=data_loader,
                  valid_data_loader=valid_data_loader,
                  lr_scheduler=lr_scheduler)
trainer.train() # 학습 시작
```

<br>

### cf) `__getitem__`

In [11]:
class Test(object):
    def __getitem__(self, items):
        print (type(items), items)

# Driver code
test = Test()
test[5]
test[5:65:5]
test['GeeksforGeeks']
test[1, 'x', 10.0]
test['a':'z':2]
test[object()]

<class 'int'> 5
<class 'slice'> slice(5, 65, 5)
<class 'str'> GeeksforGeeks
<class 'tuple'> (1, 'x', 10.0)
<class 'slice'> slice('a', 'z', 2)
<class 'object'> <object object at 0x7f54e917ecb0>


<br>

### 3.5.2 `./trainer/trainer.py`

- 실제로 학습이 시작되는 부분
- `./base/base_trainer.py` 와 함께 살펴보면 좋다.
- `_save_checkpoint`
  - 중간중간 학습 결과를 저장

<br>

### 3.5.3 `test.py`

- gradient 가 필요 없는 과정